vpr/src/pack/greedy_candidate_selector.cpp

```diff
@@ -0,0 +1,1076 @@
+                                            const Prepacker& prepacker,
+                                            const ClusterLegalizer& cluster_legalizer,
+                                            AttractionInfo& attraction_groups) {
+    auto cluster_type = cluster_legalizer.get_cluster_type(legalization_cluster_id);
+
+    /*
+     * For each cluster, we want to explore the attraction group molecules as potential
+     * candidates for the cluster a limited number of times. This limit is imposed because
+     * if the cluster belongs to a very large attraction group, we could potentially search
+     * through its attraction group molecules for a very long time.
+     * Defining a number of times to search through the attraction groups (i.e. number of
+     * attraction group pulls) determines how many times we search through the cluster's attraction
+     * group molecules for candidate molecules.
+     */
+    AttractGroupId grp_id = cluster_gain_stats.attraction_grp_id;
+    if (grp_id == AttractGroupId::INVALID()) {
+        return;
+    }
+
+    AttractionGroup& group = attraction_groups.get_attraction_group_info(grp_id);
+    std::vector<AtomBlockId> available_atoms;
+    for (AtomBlockId atom_id : group.group_atoms) {
+        const auto& atom_model = atom_netlist_.block_model(atom_id);
+        auto itr = primitive_candidate_block_types_.find(atom_model);
+        VTR_ASSERT(itr != primitive_candidate_block_types_.end());
+        const std::vector<t_logical_block_type_ptr>& candidate_types = itr->second;
+
+        //Only consider molecules that are unpacked and of the correct type
+        if (!cluster_legalizer.is_atom_clustered(atom_id)
+            && std::find(candidate_types.begin(), candidate_types.end(), cluster_type) != candidate_types.end()) {
+            available_atoms.push_back(atom_id);
+        }
+    }
+
+    int num_available_atoms = available_atoms.size();
+    if (num_available_atoms == 0) {
+        return;
+    }
+
+    if (num_available_atoms < 500) {
+        for (AtomBlockId atom_id : available_atoms) {
+            //Only consider molecules that are unpacked and of the correct type
+            t_pack_molecule* molecule = prepacker.get_atom_molecule(atom_id);
+            if (!cluster_legalizer.is_mol_clustered(molecule) &&
+                 cluster_legalizer.is_molecule_compatible(molecule, legalization_cluster_id)) {
+                add_molecule_to_pb_stats_candidates(molecule,
+                                                    cluster_gain_stats,
+                                                    packer_opts_.feasible_block_array_size,
+                                                    attraction_groups,
+                                                    atom_netlist_);
+            }
+        }
+        return;
+    }
+
+    int min = 0;
+    int max = num_available_atoms - 1;
+
+    for (int j = 0; j < 500; j++) {
+        std::random_device rd;
+        std::mt19937 gen(rd());
```

Comment:
Not sure why we're making a random number generator here -- could just use vtr::irand (faster).

Reply:
Ugh, non-deterministic random number generator.  Delete!